clang/lib/CIR/CodeGen/LoweringPrepare.cpp

```diff
@@ -865,30 +893,379 @@ void LoweringPreparePass::lowerThreeWayCmpOp(CmpThreeWayOp op) {
+        clang::CIRGen::CIRGenFunction::GuardKind::variableGuard,
+        varDecl.getRawDecl());
+
+    cgf.buildBlock(initCheckBlock);
+  }
+
+  // The semantics of dynamic initialization of variables with static or
+  // thread storage duration depends on whether they are declared at
+  // block-scope. The initialization of such variables at block-scope can be
+  // aborted with an exception and later retried (per C++20 [stmt.dcl]p4), and
+  // recursive entry to their initialization has undefined behavior (also per
+  // C++20 [stmt.dcl]p4). For such variables declared at non-block scope,
+  // exceptions lead to termination (per C++20 [except.terminate]p1), and
+  // recursive references to the variables are governed only by the lifetime
+  // rules (per C++20 [class.cdtor]p2), which means such references are
+  // perfectly fine as long as they avoid touching memory. As a result,
+  // block-scope variables must not be marked as initialized until after
+  // initialization completes (unless the mark is reverted following an
+  // exception), but non-block-scope variables must be marked prior to
+  // initialization so that recursive accesses during initialization do not
+  // restart initialization.
+
+  // Variables used when coping with thread-safe statics and exceptions.
+  if (threadsafe) {
+    auto loc = cgf.getLoc(varDecl.getLocation());
+    // Call __cxa_guard_acquire.
+    mlir::Value value = cgf.emitNounwindRuntimeCall(
+        cgf.getLoc(varDecl.getLocation()), getGuardAcquireFn(cgm, guardPtrTy),
+        guardAddr.emitRawPointer());
+
+    mlir::Block *initBlock = nullptr;
+    {
+      mlir::OpBuilder::InsertionGuard guard(builder);
+      initBlock = builder.createBlock(builder.getBlock()->getParent());
+    }
+
+    builder.createCondBr(loc, builder.createIsNotNull(loc, value), initBlock,
+                         endBlock);
+
+    // Call __cxa_guard_abort along the exceptional edge.
+    cgf.EHStack.pushCleanup<CallGuardAbort>(clang::CIRGen::EHCleanup, guard);
+
+    cgf.buildBlock(initBlock);
+  } else if (!varDecl.isLocalVarDecl()) {
+    llvm_unreachable("NYI");
+  }
+
+  // CIR: Move the initializer from the globalOp's ctor region into the current
+  // block.
+  // TODO(CIR): Once we support exceptions we'll need to walk the ctor region to
+  // change calls to invokes.
+  auto &ctorRegion = globalOp.getCtorRegion();
+  assert(!ctorRegion.empty() && "This should never be empty here.");
+  if (!ctorRegion.hasOneBlock())
+    llvm_unreachable("Multiple blocks NYI");
+  Block &block = ctorRegion.front();
+  Block *insertBlock = builder.getInsertionBlock();
+  insertBlock->getOperations().splice(insertBlock->end(), block.getOperations(),
+                                      block.begin(), std::prev(block.end()));
+  builder.setInsertionPointToEnd(insertBlock);
+
+  ctorRegion.getBlocks().clear();
+
+  if (threadsafe) {
+    // Pop the guard-abort cleanup if we pushed one.
+    cgf.PopCleanupBlock();
```

Comment:
I have some trouble understanding why can't this be handled already during CIRGen, in theory, we just need a thin layer with some hooks to lower during loweringprepare, feels like you postponed too much. I guess that's why you are trying to move things around for using CGM